### DISCUSSION TITLE
Where exactly is the code for the actual Jolt library?

### DISCUSSION THREAD

**Discussion by ValorZard** [godot-jolt/godot-jolt#61] [Questions] [answered]
taking a quick look through this thing, and I can't seem to find any references to the Jolt physics library itself, nor do i see a git submodule or something like that. Has the Jolt physics library codebase not been integrated yet?

**Comment by mihe** [comment, marked as answer]
Anything prefixed with the `JPH::` namespace is referencing Jolt.

[`JoltSpace3D`](https://github.com/godot-jolt/godot-jolt/blob/878ba22126428664b8a6f751fb62e0b1d597cf87/src/jolt_space_3d.cpp#L28-L36) is what's wrapping Jolt's `JPH::PhysicsSystem`. I guess you could say it starts there. Bodies are currently created in [`JoltSpace3D::create_object`](https://github.com/godot-jolt/godot-jolt/blob/878ba22126428664b8a6f751fb62e0b1d597cf87/src/jolt_space_3d.cpp#L136-L210) and shapes are created in [`jolt_shape_3d.cpp`](https://github.com/godot-jolt/godot-jolt/blob/878ba22126428664b8a6f751fb62e0b1d597cf87/src/jolt_shape_3d.cpp), just to name a few things.

I don't use Git submodules. I use CMake's built-in Git support to deal with all that, using their [`ExternalProject`](https://cmake.org/cmake/help/v3.22/module/ExternalProject.html) module, which I've wrapped in my own abstraction to improve the ergonomics a bit.

All it really means is that I can deal with third-party dependencies much like I would have with Git submodules, like being able to reference an external Git repository and a SHA hash, but you don't have to append `--recursive` when cloning and all that.

As soon as you start a build it'll download the third-party libraries as-needed, and it'll update them as-needed when the commit hash changes.

You can find the reference to Jolt's Git repository/commit in [`cmake/GodotJoltExternalJolt.cmake`](https://github.com/godot-jolt/godot-jolt/blob/878ba22126428664b8a6f751fb62e0b1d597cf87/cmake/GodotJoltExternalJolt.cmake#L32-L33).

I should probably have mentioned a lot of this in `docs/hacking.md`. I'll see about adding that.
